HDDS-11041. Add admin request filter for S3 requests and UGI support for GrpcOmTransport

devabhishekpal opened the pull request:
## What changes were proposed in this pull request?

HDDS-11041. Add admin request filter for S3 requests and UGI support for GrpcOmTransport

Please describe your PR in detail:
* Currently the s3secret endpoints do not support checking for the user making the request. This causes issues like [HDDS-11040](https://issues.apache.org/jira/browse/HDDS-11040) where any user can modify the secrets
* As a part of this PR we are adding a new filter to check if the user making the request is a part of Ozone Admins, if not we will not trigger the generation/revocation of secret
* We are also adding the passing of UserGroupInformation to the GrpcOmTransport while creating the proxy connection to allow accessing the user ticket similar to the Hadoop3OmTransport

## What is the link to the Apache JIRA

https://issues.apache.org/jira/browse/HDDS-11041

## How was this patch tested?
Patch was tested using unit tests

adoroszlai commented:
@ivanzlenko please take a look

devabhishekpal commented on hadoop-ozone/s3gateway/src/main/java/org/apache/hadoop/ozone/s3secret/S3SecretAdminFilter.java:
```diff
@@ -0,0 +1,80 @@
+   * @param user Stores the user to filter
+   * @return true if the user is admin else false
+   */
+  private boolean isAdmin(UserGroupInformation user) {
```
Thanks for pointing this out @myskov.
Updated the code to reflect this change, and also changed the getter methods on OzoneConfigUtil to public to allow access outside package.

devabhishekpal replied:
Addressed in [d5272f5](https://github.com/apache/ozone/pull/7268/commits/d5272f54043a6fa156f3d2ac5fb1edb6a5a364e4)

devabhishekpal commented:
Hi @ivanzlenko, new tests have not yet been added. We will need to modify existing tests to  add user as an admin and generate request.
New test will be having a random user try to generate the request and get a forbidden response.

Waiting for go on code change before modifying the tests.

devabhishekpal commented:
Yes, these new cases will be added to check for the filter working

devabhishekpal commented on hadoop-ozone/s3gateway/src/main/java/org/apache/hadoop/ozone/s3secret/S3SecretAdminFilter.java:
```diff
@@ -0,0 +1,87 @@
+import java.util.Collections;
+
+/**
+ * Filter that disables all endpoints annotated with {@link S3AdminEndpoint}.
```
Yup, sorry I copied over the previous doc for S3SecretEnabled annotation.
Modified in [932fffc](https://github.com/apache/ozone/pull/7268/commits/932fffc138ac2681f1362b586b1781f4280bd7b9)

devabhishekpal commented on hadoop-ozone/common/src/main/java/org/apache/hadoop/ozone/om/ha/GrpcOMFailoverProxyProvider.java:
```diff
@@ -118,9 +125,21 @@ private T createOMProxy() throws IOException {
+    return (T) RPC.getProtocolProxy(
+      getInterface(),
+      0,
+      addr, ugi,
```
Addressed in [932fffc](https://github.com/apache/ozone/pull/7268/commits/932fffc138ac2681f1362b586b1781f4280bd7b9)

devabhishekpal commented on hadoop-ozone/common/src/main/java/org/apache/hadoop/ozone/om/ha/GrpcOMFailoverProxyProvider.java:
```diff
@@ -118,9 +125,21 @@ private T createOMProxy() throws IOException {
+      connectionRetryPolicy).getProxy();
   }
 
+
```
Addressed in [932fffc](https://github.com/apache/ozone/pull/7268/commits/932fffc138ac2681f1362b586b1781f4280bd7b9)

devabhishekpal commented on hadoop-ozone/s3gateway/src/main/java/org/apache/hadoop/ozone/s3secret/S3SecretAdminFilter.java:
```diff
@@ -0,0 +1,87 @@
+ *   - ozone.administrators
+ *   - ozone.administrators.groups
+ */
+
```
Addressed in [932fffc](https://github.com/apache/ozone/pull/7268/commits/932fffc138ac2681f1362b586b1781f4280bd7b9)

devabhishekpal commented on hadoop-ozone/s3gateway/src/main/java/org/apache/hadoop/ozone/s3secret/S3SecretAdminFilter.java:
```diff
@@ -0,0 +1,87 @@
+/**
+ * Filter that disables all endpoints annotated with {@link S3AdminEndpoint}.
+ * Condition is based on the value of the configuration keys for
+ *   - ozone.administrators
```
Addressed in [932fffc](https://github.com/apache/ozone/pull/7268/commits/932fffc138ac2681f1362b586b1781f4280bd7b9)

devabhishekpal commented on hadoop-ozone/s3gateway/src/main/java/org/apache/hadoop/ozone/s3secret/S3SecretAdminFilter.java:
```diff
@@ -0,0 +1,87 @@
+    final Principal userPrincipal = requestContext.getSecurityContext().getUserPrincipal();
+    if (null != userPrincipal) {
+      UserGroupInformation user =
+          UserGroupInformation.createRemoteUser(userPrincipal.getName());
```
Addressed in [932fffc](https://github.com/apache/ozone/pull/7268/commits/932fffc138ac2681f1362b586b1781f4280bd7b9)

devabhishekpal commented on hadoop-ozone/s3gateway/src/main/java/org/apache/hadoop/ozone/s3secret/S3AdminEndpoint.java:
```diff
@@ -0,0 +1,34 @@
+import javax.ws.rs.NameBinding;
+
+/**
+ * Annotation to disable S3 Secure Endpoint.
```
Addressed in [932fffc](https://github.com/apache/ozone/pull/7268/commits/932fffc138ac2681f1362b586b1781f4280bd7b9)

devabhishekpal commented on hadoop-ozone/s3gateway/src/main/java/org/apache/hadoop/ozone/s3secret/S3SecretAdminFilter.java:
```diff
@@ -0,0 +1,87 @@
+
+  @Override
+  public void filter(ContainerRequestContext requestContext)
+      throws IOException {
```
Addressed in [932fffc](https://github.com/apache/ozone/pull/7268/commits/932fffc138ac2681f1362b586b1781f4280bd7b9)

sumitagrawl commented on hadoop-ozone/s3gateway/src/main/java/org/apache/hadoop/ozone/s3secret/S3SecretAdminFilter.java:
```diff
@@ -0,0 +1,87 @@
+    try {
+      s3Admins = OzoneConfigUtil.getS3AdminsFromConfig(conf);
+    } catch (IOException ie) {
+      // We caught an exception while trying to log in using the UGI user
```
We Can not call OzoneManager.isS3Admin, as its another instance not supposed to be initialized for S3G.
But if required,
1. can get object in common method for both OzoneManager and S3g,

devabhishekpal replied:
Addressed in [a0ac299](https://github.com/apache/ozone/pull/7268/commits/a0ac299ee1c8e2085ecbfe1c3fc38b8cc4396c8f).

Created OmUserUtils class to later on be able to add more utils as required for Om user related utils.
The same is called in both OzoneManager and S3AdminFilter in order to fetch the users.

devabhishekpal commented on hadoop-ozone/s3gateway/src/main/java/org/apache/hadoop/ozone/s3secret/S3SecretAdminFilter.java:
```diff
@@ -0,0 +1,87 @@
+import javax.inject.Inject;
+import javax.ws.rs.container.ContainerRequestContext;
+import javax.ws.rs.container.ContainerRequestFilter;
+import javax.ws.rs.core.Response;
```
Addressed in [a0ac299](https://github.com/apache/ozone/pull/7268/commits/a0ac299ee1c8e2085ecbfe1c3fc38b8cc4396c8f)

devabhishekpal commented on hadoop-ozone/s3gateway/src/main/java/org/apache/hadoop/ozone/s3secret/S3SecretAdminFilter.java:
```diff
@@ -0,0 +1,87 @@
+      UserGroupInformation user = UserGroupInformation.createRemoteUser(userPrincipal.getName());
+      if (!isAdmin(user)) {
+        requestContext.abortWith(Response.status(403)
+            .entity("Non-Admin user accessing endpoint")
```
Addressed in [a0ac299](https://github.com/apache/ozone/pull/7268/commits/a0ac299ee1c8e2085ecbfe1c3fc38b8cc4396c8f)

devabhishekpal commented on hadoop-ozone/s3gateway/src/main/java/org/apache/hadoop/ozone/s3secret/S3SecretAdminFilter.java:
```diff
@@ -0,0 +1,87 @@
+      UserGroupInformation user =
+          UserGroupInformation.createRemoteUser(userPrincipal.getName());
+      if (!isAdmin(user)) {
+        requestContext.abortWith(Response.status(Response.Status.FORBIDDEN)
```
Addressed in [a0ac299](https://github.com/apache/ozone/pull/7268/commits/a0ac299ee1c8e2085ecbfe1c3fc38b8cc4396c8f)

devabhishekpal commented on hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/common/OmUserUtils.java:
```diff
@@ -0,0 +1,87 @@
+ * Utility class to store User related utilities.
+ */
+public final class OmUserUtils {
+  private static final Logger LOG =
```
Addressed in [ff14d26](https://github.com/apache/ozone/pull/7268/commits/ff14d265465018ccd15ebc1fb727921d871ad8a9)

devabhishekpal commented on hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/common/OmUserUtils.java:
```diff
@@ -0,0 +1,87 @@
+    } catch (IOException ie) {
+      s3Admins = null;
+    }
+    Collection<String> s3AdminGroups =
```
Addressed in [ff14d26](https://github.com/apache/ozone/pull/7268/commits/ff14d265465018ccd15ebc1fb727921d871ad8a9)

devabhishekpal commented on hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/common/OmUserUtils.java:
```diff
@@ -0,0 +1,87 @@
+
+  /**
+   * Check if the provided user is a part of the S3 admins.
+   * @param user Stores the user to verify
```
Addressed in [ff14d26](https://github.com/apache/ozone/pull/7268/commits/ff14d265465018ccd15ebc1fb727921d871ad8a9)

devabhishekpal commented on hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/common/OmUserUtils.java:
```diff
@@ -0,0 +1,87 @@
+  public static boolean isS3Admin(UserGroupInformation user,
+                                  OzoneConfiguration conf) {
+    OzoneAdmins s3Admins = getS3Admins(conf);
+    return null != user && s3Admins.isAdmin(user);
```
Addressed in [ff14d26](https://github.com/apache/ozone/pull/7268/commits/ff14d265465018ccd15ebc1fb727921d871ad8a9)

devabhishekpal commented on hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/common/OmUserUtils.java:
```diff
@@ -0,0 +1,87 @@
+  }
+
+  /**
+   * Check if the provided user is a part of the S3 admins.
```
Addressed in [ff14d26](https://github.com/apache/ozone/pull/7268/commits/ff14d265465018ccd15ebc1fb727921d871ad8a9)

devabhishekpal commented on hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/common/OmUserUtils.java:
```diff
@@ -0,0 +1,87 @@
+
+  /**
+   * Check if the provided user is a part of the S3 admins.
+   * @param user Stores the user to verify
```
Addressed in [ff14d26](https://github.com/apache/ozone/pull/7268/commits/ff14d265465018ccd15ebc1fb727921d871ad8a9)

devabhishekpal commented on hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/common/OmUserUtils.java:
```diff
@@ -0,0 +1,87 @@
+   * Check if the provided user is a part of the S3 admins.
+   * @param user Stores the user to verify
+   * @param s3Admins Stores the users and groups that are admins
+   * @return true if the provided user is an S3 admin else false
```
Addressed in [ff14d26](https://github.com/apache/ozone/pull/7268/commits/ff14d265465018ccd15ebc1fb727921d871ad8a9)

devabhishekpal commented on hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/common/OmUserUtils.java:
```diff
@@ -0,0 +1,87 @@
+  /**
+   * Check if the provided user is a part of the S3 admins.
+   * @param user Stores the user to verify
+   * @param s3Admins Stores the users and groups that are admins
```
Addressed in [ff14d26](https://github.com/apache/ozone/pull/7268/commits/ff14d265465018ccd15ebc1fb727921d871ad8a9)

devabhishekpal commented on hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/common/OmUserUtils.java:
```diff
@@ -0,0 +1,87 @@
+   * @param conf Stores the Ozone configuration being used
+   * @return true if the provided user is an S3 admin else false
+   */
+  public static boolean isS3Admin(UserGroupInformation user,
```
Addressed in [ff14d26](https://github.com/apache/ozone/pull/7268/commits/ff14d265465018ccd15ebc1fb727921d871ad8a9)

devabhishekpal commented on hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/common/OmUserUtils.java:
```diff
@@ -0,0 +1,87 @@
+   * Check if the provided user is a part of the S3 admins.
+   * @param user Stores the user to verify
+   * @param conf Stores the Ozone configuration being used
+   * @return true if the provided user is an S3 admin else false
```
Addressed in [ff14d26](https://github.com/apache/ozone/pull/7268/commits/ff14d265465018ccd15ebc1fb727921d871ad8a9)

adoroszlai commented on hadoop-ozone/s3gateway/src/test/java/org/apache/hadoop/ozone/s3secret/TestSecretGenerate.java:
```diff
@@ -51,6 +52,7 @@
  */
 @ExtendWith(MockitoExtension.class)
 class TestSecretGenerate {
```
https://github.com/devabhishekpal/ozone/actions/runs/11329186708/job/31504841331#step:6:2919

adoroszlai commented on hadoop-ozone/s3gateway/src/test/java/org/apache/hadoop/ozone/s3secret/TestSecretRevoke.java:
```diff
@@ -26,42 +26,43 @@
  * Test for S3 secret revoke endpoint.
  */
 @ExtendWith(MockitoExtension.class)
 public class TestSecretRevoke {
```
https://github.com/devabhishekpal/ozone/actions/runs/11329186708/job/31504841331#step:6:2864

devabhishekpal commented on hadoop-ozone/common/src/main/java/org/apache/hadoop/ozone/conf/OzoneS3ConfigUtils.java:
```diff
@@ -0,0 +1,133 @@
+    if (ozAdmins == null || ozAdmins.isEmpty()) {
+      ozAdmins = conf.getTrimmedStringCollection(OZONE_ADMINISTRATORS);
+    }
+    String omSPN = UserGroupInformation.getCurrentUser().getShortUserName();
```
This is also present in the Ozone Admin class when trying to get the OzoneAdmin from a config, we pass a starter user which is set as the admin in case the provided user is not already present as an admin

devabhishekpal commented on hadoop-ozone/common/src/main/java/org/apache/hadoop/ozone/conf/OzoneS3ConfigUtils.java:
```diff
@@ -0,0 +1,133 @@
+ * Config based utilities for Ozone S3.
+ */
+public final class OzoneS3ConfigUtils {
+  static final Logger LOG = LoggerFactory.getLogger(OzoneS3ConfigUtils.class);
```
Addressed in [2daecc9](https://github.com/apache/ozone/pull/7268/commits/2daecc9cf71112d0c0813479e3d49181cf230e9b)

devabhishekpal commented on hadoop-ozone/common/src/main/java/org/apache/hadoop/ozone/conf/OzoneS3ConfigUtils.java:
```diff
@@ -0,0 +1,133 @@
+   * defaults to ozone.administrators value.
+   */
+  public static Collection<String> getS3AdminsFromConfig(OzoneConfiguration conf) throws IOException {
+    Collection<String> ozAdmins = conf.getTrimmedStringCollection(OZONE_S3_ADMINISTRATORS);
```
Addressed in [2daecc9](https://github.com/apache/ozone/pull/7268/commits/2daecc9cf71112d0c0813479e3d49181cf230e9b)

devabhishekpal commented on hadoop-ozone/common/src/main/java/org/apache/hadoop/ozone/conf/OzoneS3ConfigUtils.java:
```diff
@@ -0,0 +1,133 @@
+   *                 of the S3 administrator users and groups in the system
+   * @return {@code true} if the provided user is an S3 administrator else {@code false}
+   */
+  public static boolean isS3Admin(@Nullable UserGroupInformation user, OzoneAdmins s3Admins) {
```
Removed in [2daecc9](https://github.com/apache/ozone/pull/7268/commits/2daecc9cf71112d0c0813479e3d49181cf230e9b)

devabhishekpal replied:
Reverted this as this is being used in OzoneManager and is added as a part of reducing code duplication.
Refer to: https://github.com/apache/ozone/blob/8568075ddb75b4a6785cdb31b00b4a592058dc78/hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/OzoneManager.java#L4347 for this refactor source

devabhishekpal commented on hadoop-ozone/common/src/test/java/org/apache/hadoop/ozone/conf/TestOzoneS3ConfigUtils.java:
```diff
@@ -0,0 +1,72 @@
+public class TestOzoneS3ConfigUtils {
+
+  @Test
+  public void testS3AdminExtraction() throws IOException {
```
Addressed in [2daecc9](https://github.com/apache/ozone/pull/7268/commits/2daecc9cf71112d0c0813479e3d49181cf230e9b)

devabhishekpal commented on hadoop-ozone/common/src/test/java/org/apache/hadoop/ozone/conf/TestOzoneS3ConfigUtils.java:
```diff
@@ -0,0 +1,72 @@
+  }
+
+  @Test
+  public void testS3AdminGroupExtraction() {
```
Addressed in [2daecc9](https://github.com/apache/ozone/pull/7268/commits/2daecc9cf71112d0c0813479e3d49181cf230e9b)

devabhishekpal commented on hadoop-ozone/common/src/test/java/org/apache/hadoop/ozone/conf/TestOzoneS3ConfigUtils.java:
```diff
@@ -0,0 +1,72 @@
+/**
+ * This class is to test S3 configuration based utils.
+ */
+public class TestOzoneS3ConfigUtils {
```
Addressed in [2daecc9](https://github.com/apache/ozone/pull/7268/commits/2daecc9cf71112d0c0813479e3d49181cf230e9b)

devabhishekpal commented on hadoop-ozone/common/src/test/java/org/apache/hadoop/ozone/conf/TestOzoneS3ConfigUtils.java:
```diff
@@ -0,0 +1,72 @@
+  }
+
+  @Test
+  public void testS3AdminExtractionWithFallback() throws IOException {
```
Addressed in [2daecc9](https://github.com/apache/ozone/pull/7268/commits/2daecc9cf71112d0c0813479e3d49181cf230e9b)

devabhishekpal commented on hadoop-ozone/common/src/test/java/org/apache/hadoop/ozone/conf/TestOzoneS3ConfigUtils.java:
```diff
@@ -0,0 +1,72 @@
+  }
+
+  @Test
+  public void testS3AdminGroupExtractionWithFallback() {
```
Addressed in [2daecc9](https://github.com/apache/ozone/pull/7268/commits/2daecc9cf71112d0c0813479e3d49181cf230e9b)

devabhishekpal commented on hadoop-ozone/common/src/main/java/org/apache/hadoop/ozone/conf/OzoneS3ConfigUtils.java:
```diff
@@ -0,0 +1,133 @@
+   * @param conf An instance of {@link OzoneConfiguration} being used
+   * @return {@code true} if the provided user is an S3 administrator else {@code false}
+   */
+  public static boolean isS3Admin(@Nullable UserGroupInformation user, OzoneConfiguration conf) {
```
Addressed in [2daecc9](https://github.com/apache/ozone/pull/7268/commits/2daecc9cf71112d0c0813479e3d49181cf230e9b)

devabhishekpal commented on hadoop-ozone/common/src/main/java/org/apache/hadoop/ozone/conf/OzoneS3ConfigUtils.java:
```diff
@@ -0,0 +1,133 @@
+   * @param conf  Stores an instance of {@link OzoneConfiguration} being used
+   * @return an instance of {@link OzoneAdmins} containing the S3 admin users and groups
+   */
+  public static OzoneAdmins getS3Admins(OzoneConfiguration conf) {
```
Addressed in [2daecc9](https://github.com/apache/ozone/pull/7268/commits/2daecc9cf71112d0c0813479e3d49181cf230e9b)

devabhishekpal commented:
Hi @ivanzlenko, for the new filter it seems if we are adding tests to the `TestSecret*` testset, the filter is not triggered as the methods are being called directly and the filter doesn't take effect.
We can go ahead with tests just for the filter like explicitly passing a mocked requestContext with invalid user and asserting that FORBIDDEN is thrown.

But that doesn't seem to be much use. Do you have any inputs regarding this?
I tried manually testing it via setting up a `ozonesecure` docker environment, but we cannot send requests via POSTMAN as the kerberos auth is not present in the local machine and docker is isolated.
It keeps giving 401 error as the user itself is unauthorized and doesn't reach the admin check step.
```2024-10-21 15:14:07 2024-10-21 09:44:07,266 [qtp1078566479-63] WARN server.AuthenticationFilter: AuthenticationToken ignored: Unauthorized access
```

Inputs would be great in this regard @ivanzlenko @adoroszlai @myskov

adoroszlai reviewed (changes requested):
Thanks @devabhishekpal for working on this.

1. Please enable existing robot tests related to HDDS-11041, e.g.:

https://github.com/apache/ozone/blob/8568075ddb75b4a6785cdb31b00b4a592058dc78/hadoop-ozone/dist/src/main/smoketest/s3/secretgenerate.robot#L48

2. Please add a new test case similar to `... By Username For Other User`, but with `testuser2` (non-admin) logged in, and expecting request to be rejected.

devabhishekpal commented on hadoop-ozone/common/pom.xml:
```diff
@@ -65,6 +65,10 @@ https://maven.apache.org/xsd/maven-4.0.0.xsd">
+    <dependency>
+      <groupId>org.apache.ozone</groupId>
+      <artifactId>hdds-server-framework</artifactId>
+    </dependency>
```
Addressed in [bf6fb02](https://github.com/apache/ozone/pull/7268/commits/bf6fb0239af226cebc857632e31b968b635d64b3)

devabhishekpal commented:
Hi @ivanzlenko, @adoroszlai, @myskov, @vtutrinov could you take another look at the changes?
Thanks.

- Added new robot test to verify the filter
- Refactored the common config related utils to OzoneAdmins class
- Re-enabled the skipped robot tests added by @ivanzlenko initially as a part of HDDS-11040
- Added unit tests for the new utils

adoroszlai commented on hadoop-hdds/framework/src/test/java/org/apache/hadoop/hdds/server/TestOzoneAdminUtils.java:
```diff
@@ -0,0 +1,158 @@
+/**
+ * This class is to test the utilities present in the OzoneAdmins class.
+ */
+class TestOzoneAdminUtils {
```
nit: should be `TestOzoneAdmins`

devabhishekpal replied:
Addressed in [2acffda](https://github.com/apache/ozone/pull/7268/commits/2acffdaa039b0d8ef0b54f81d146664e574e8a96)

adoroszlai commented on hadoop-ozone/common/src/main/java/org/apache/hadoop/ozone/om/ha/GrpcOMFailoverProxyProvider.java:
```diff
@@ -116,9 +126,35 @@ protected void loadOMClientConfigs(ConfigurationSource config, String omSvcId)
+   * @throws IOException if any error occurs while trying to get the proxy
+   */
+  private T createOmProxy(InetSocketAddress address) throws IOException {
     Configuration hadoopConf =
         LegacyHadoopConfigurationSource.asHadoopConfiguration(getConf());
-    return (T) RPC.getProxy(getInterface(), 0, addr, hadoopConf);
+
+    // TODO: Post upgrade to Protobuf 3.x we need to use ProtobufRpcEngine2
+    RPC.setProtocolEngine(hadoopConf, getInterface(), ProtobufRpcEngine.class);
+
+    // Ensure we do not attempt retry on the same OM in case of exceptions
+    RetryPolicy connectionRetryPolicy = RetryPolicies.failoverOnNetworkException(0);
+
+    return (T) RPC.getProtocolProxy(
+        getInterface(),
+        protocolVer,
+        address,
+        ugi,
+        hadoopConf,
+        NetUtils.getDefaultSocketFactory(hadoopConf),
+        (int) OmUtils.getOMClientRpcTimeOut(getConf()),
+        connectionRetryPolicy
+    ).getProxy();
```
This seems to be copied from `HadoopRpcOMFailoverProxyProvider`.  Instead of duplicating it, can it be pushed to the parent class `OMFailoverProxyProviderBase`?

devabhishekpal replied:
Addressed in [2acffda](https://github.com/apache/ozone/pull/7268/commits/2acffdaa039b0d8ef0b54f81d146664e574e8a96)

adoroszlai replied:
Thanks.  Can you please check unused imports?

```
hadoop-ozone/common/src/main/java/org/apache/hadoop/ozone/om/ha/HadoopRpcOMFailoverProxyProvider.java
 32: Unused import - org.apache.hadoop.conf.Configuration.
 34: Unused import - org.apache.hadoop.hdds.utils.LegacyHadoopConfigurationSource.
 36: Unused import - org.apache.hadoop.io.retry.RetryPolicies.
 37: Unused import - org.apache.hadoop.io.retry.RetryPolicy.
 38: Unused import - org.apache.hadoop.ipc.ProtobufRpcEngine.
 40: Unused import - org.apache.hadoop.net.NetUtils.
hadoop-ozone/common/src/main/java/org/apache/hadoop/ozone/om/ha/GrpcOMFailoverProxyProvider.java
 21: Unused import - org.apache.hadoop.conf.Configuration.
 25: Unused import - org.apache.hadoop.hdds.utils.LegacyHadoopConfigurationSource.
 26: Unused import - org.apache.hadoop.io.retry.RetryPolicies.
 27: Unused import - org.apache.hadoop.io.retry.RetryPolicy.
 28: Unused import - org.apache.hadoop.ipc.ProtobufRpcEngine.
```

adoroszlai reviewed (commented):
Thanks @devabhishekpal for updating the patch.  Mostly looks good.

I just realized that `GrpcOmTransport` is not used in any acceptance test environments.  Can you please enable it in `ozonesecure` env. by the following change?

```diff
diff --git hadoop-ozone/dist/src/main/compose/ozonesecure/docker-compose.yaml hadoop-ozone/dist/src/main/compose/ozonesecure/docker-compose.yaml
index 39d26c362f..3fb7525f20 100644
--- hadoop-ozone/dist/src/main/compose/ozonesecure/docker-compose.yaml
+++ hadoop-ozone/dist/src/main/compose/ozonesecure/docker-compose.yaml
@@ -96,7 +96,7 @@ services:
       - 9878:9878
     env_file:
       - ./docker-config
-    command: ["/opt/hadoop/bin/ozone","s3g"]
+    command: ["/opt/hadoop/bin/ozone","s3g","-Dozone.om.transport.class=org.apache.hadoop.ozone.om.protocolPB.GrpcOmTransportFactory"]
     environment:
       OZONE_OPTS:
   recon:
```

adoroszlai commented on hadoop-ozone/common/src/main/java/org/apache/hadoop/ozone/om/ha/GrpcOMFailoverProxyProvider.java:
```diff
@@ -59,10 +63,14 @@ public class GrpcOMFailoverProxyProvider<T> extends
                                      Class<T> protocol) throws IOException {
-    super(configuration, omServiceId, protocol);
+    super(configuration, ugi, omServiceId, protocol);
+    this.protocolVer = RPC.getProtocolVersion(protocol);
```
`protocolVer` is unused.

devabhishekpal replied:
Addressed in [c52deae](https://github.com/apache/ozone/pull/7268/commits/c52deaec7d51ec9adea755d75d83e82bcdd74e3d)

adoroszlai commented on hadoop-ozone/common/src/main/java/org/apache/hadoop/ozone/om/ha/HadoopRpcOMFailoverProxyProvider.java:
```diff
@@ -75,8 +74,7 @@ public HadoopRpcOMFailoverProxyProvider(ConfigurationSource configuration,
-    super(configuration, omServiceId, protocol);
-    this.ugi = ugi;
+    super(configuration, ugi, omServiceId, protocol);
     this.omVersion = RPC.getProtocolVersion(protocol);
```
`omVersion` is no longer used.

devabhishekpal replied:
Addressed in [c52deae](https://github.com/apache/ozone/pull/7268/commits/c52deaec7d51ec9adea755d75d83e82bcdd74e3d)

devabhishekpal commented:
Thanks for the review @adoroszlai , addressed the test failures.

adoroszlai reviewed (approved):
Thanks @devabhishekpal for iterating on the patch.

devabhishekpal commented:
Thanks for the input @ivanzlenko.
We re-enabled robot tests to verify this as well as new robot tests to verify the admin filter.
New unit tests were also added for the utils. It would be great if you could take a look

devabhishekpal commented on hadoop-hdds/framework/src/test/java/org/apache/hadoop/hdds/server/TestOzoneAdmins.java:
```diff
@@ -0,0 +1,158 @@
+/**
+ * This class is to test the utilities present in the OzoneAdmins class.
+ */
+class TestOzoneAdmins {
```
BeforeEach wouldn't be much use here since the configs are changed for different test sets.
Instead parameterised the tests to avoid test code duplication

devabhishekpal replied:
Addressed in [9e8c64e](https://github.com/apache/ozone/pull/7268/commits/9e8c64e08e7b902e8a7331d5654c877ad43b97e4) could you take a look if it is okay?

devabhishekpal commented on hadoop-hdds/framework/src/main/java/org/apache/hadoop/hdds/server/OzoneAdmins.java:
```diff
@@ -186,4 +191,85 @@ public static Collection<String> getOzoneReadOnlyAdminsGroupsFromConfig(
   }
+
+  /**
+   * Get the list of S3 administrators from Ozone config.
```
Addressed in [9e8c64e](https://github.com/apache/ozone/pull/7268/commits/9e8c64e08e7b902e8a7331d5654c877ad43b97e4)

devabhishekpal commented on hadoop-hdds/framework/src/main/java/org/apache/hadoop/hdds/server/OzoneAdmins.java:
```diff
@@ -186,4 +191,85 @@ public static Collection<String> getOzoneReadOnlyAdminsGroupsFromConfig(
+   * If ozone.s3.administrators.groups value is empty or unset,
+   * defaults to the ozone.administrators.groups value
+   */
+  public static Collection<String> getS3AdminsGroupsFromConfig(OzoneConfiguration conf) {
```
Addressed in [9e8c64e](https://github.com/apache/ozone/pull/7268/commits/9e8c64e08e7b902e8a7331d5654c877ad43b97e4)

devabhishekpal commented on hadoop-hdds/framework/src/main/java/org/apache/hadoop/hdds/server/OzoneAdmins.java:
```diff
@@ -186,4 +191,85 @@ public static Collection<String> getOzoneReadOnlyAdminsGroupsFromConfig(
+    try {
+      s3Admins = getS3AdminsFromConfig(conf);
+    } catch (IOException ie) {
+      s3Admins = null;
```
Addressed in [9e8c64e](https://github.com/apache/ozone/pull/7268/commits/9e8c64e08e7b902e8a7331d5654c877ad43b97e4)

devabhishekpal commented on hadoop-hdds/framework/src/test/java/org/apache/hadoop/hdds/server/TestOzoneAdmins.java:
```diff
@@ -0,0 +1,158 @@
+  void testS3AdminGroupExtractionWithFallback() {
+    OzoneConfiguration configuration = new OzoneConfiguration();
+    configuration.set(OzoneConfigKeys.OZONE_ADMINISTRATORS_GROUPS,
+        "test1, test2");
```
Addressed in [9e8c64e](https://github.com/apache/ozone/pull/7268/commits/9e8c64e08e7b902e8a7331d5654c877ad43b97e4)

devabhishekpal commented on hadoop-ozone/common/src/main/java/org/apache/hadoop/ozone/om/ha/GrpcOMFailoverProxyProvider.java:
```diff
@@ -59,10 +57,12 @@ public class GrpcOMFailoverProxyProvider<T> extends
   public static final Logger LOG =
       LoggerFactory.getLogger(GrpcOMFailoverProxyProvider.class);
 
+
```
Addressed in [9e8c64e](https://github.com/apache/ozone/pull/7268/commits/9e8c64e08e7b902e8a7331d5654c877ad43b97e4)

devabhishekpal commented on hadoop-hdds/framework/src/test/java/org/apache/hadoop/hdds/server/TestOzoneAdmins.java:
```diff
@@ -0,0 +1,158 @@
+  void testS3AdminGroupExtraction() {
+    OzoneConfiguration configuration = new OzoneConfiguration();
+    configuration.set(OzoneConfigKeys.OZONE_S3_ADMINISTRATORS_GROUPS,
+        "test1, test2");
```
Addressed in [9e8c64e](https://github.com/apache/ozone/pull/7268/commits/9e8c64e08e7b902e8a7331d5654c877ad43b97e4)

devabhishekpal commented on hadoop-hdds/framework/src/main/java/org/apache/hadoop/hdds/server/OzoneAdmins.java:
```diff
@@ -186,4 +191,85 @@ public static Collection<String> getOzoneReadOnlyAdminsGroupsFromConfig(
+  }
+
+  /**
+   * Get the list of the groups that are a part S3 administrators from Ozone config.
```
Addressed in [9e8c64e](https://github.com/apache/ozone/pull/7268/commits/9e8c64e08e7b902e8a7331d5654c877ad43b97e4)

devabhishekpal commented on hadoop-hdds/framework/src/main/java/org/apache/hadoop/hdds/server/OzoneAdmins.java:
```diff
@@ -186,4 +191,85 @@ public static Collection<String> getOzoneReadOnlyAdminsGroupsFromConfig(
+   * @return an instance of {@link OzoneAdmins} containing the S3 admin users and groups
+   */
+  public static OzoneAdmins getS3Admins(OzoneConfiguration conf) {
+    Collection<String> s3Admins;
```
Addressed in [9e8c64e](https://github.com/apache/ozone/pull/7268/commits/9e8c64e08e7b902e8a7331d5654c877ad43b97e4)

devabhishekpal commented on hadoop-hdds/framework/src/main/java/org/apache/hadoop/hdds/server/OzoneAdmins.java:
```diff
@@ -186,4 +192,91 @@ public static Collection<String> getOzoneReadOnlyAdminsGroupsFromConfig(
+    }
+
+    String omSPN = UserGroupInformation.getCurrentUser().getShortUserName();
+    if (!ozoneAdmins.contains(omSPN)) {
```
Thanks for pointing this out, addressed in [e569e5f](https://github.com/apache/ozone/pull/7268/commits/e569e5f31b92e6426b3797d0a436457fe59b3da0)

devabhishekpal commented on hadoop-hdds/framework/src/test/java/org/apache/hadoop/hdds/server/TestOzoneAdmins.java:
```diff
@@ -0,0 +1,122 @@
+/**
+ * This class is to test the utilities present in the OzoneAdmins class.
+ */
+class TestOzoneAdmins {
```
Addressed in [e569e5f](https://github.com/apache/ozone/pull/7268/commits/e569e5f31b92e6426b3797d0a436457fe59b3da0)

devabhishekpal commented on hadoop-hdds/framework/src/main/java/org/apache/hadoop/hdds/server/OzoneAdmins.java:
```diff
@@ -186,4 +192,91 @@ public static Collection<String> getOzoneReadOnlyAdminsGroupsFromConfig(
+   * </ul>
+   * @param conf An instance of {@link OzoneConfiguration} being used
+   * @return A {@link Collection} of the S3 administrator users
+   *
```
Addressed in [e569e5f](https://github.com/apache/ozone/pull/7268/commits/e569e5f31b92e6426b3797d0a436457fe59b3da0)

devabhishekpal commented on hadoop-hdds/framework/src/main/java/org/apache/hadoop/hdds/server/OzoneAdmins.java:
```diff
@@ -33,6 +36,9 @@
 import static org.apache.hadoop.ozone.OzoneConfigKeys.OZONE_READONLY_ADMINISTRATORS_GROUPS;
+import static org.apache.hadoop.ozone.OzoneConfigKeys.OZONE_S3_ADMINISTRATORS;
+import static org.apache.hadoop.ozone.OzoneConfigKeys.OZONE_S3_ADMINISTRATORS_GROUPS;
+
```
Addressed in [e569e5f](https://github.com/apache/ozone/pull/7268/commits/e569e5f31b92e6426b3797d0a436457fe59b3da0)

devabhishekpal commented:
Thanks for the quick fix @adoroszlai

devabhishekpal commented:
Thanks for the reviews @ivanzlenko @adoroszlai @myskov @vtutrinov @sumitagrawl

adoroszlai commented:
Thanks @devabhishekpal for the patch, @ivanzlenko, @myskov, @sumitagrawl, @vtutrinov for the reviews.
